nifi-nar-bundles/nifi-extension-utils/nifi-dbcp-base/src/main/java/org/apache/nifi/dbcp/AbstractDBCPConnectionPool.java

```diff
@@ -365,168 +121,151 @@ public List<ConfigVerificationResult> verify(final ConfigurationContext context,
-        final String passw = context.getProperty(DB_PASSWORD).evaluateAttributeExpressions().getValue();
-        final Integer maxTotal = context.getProperty(MAX_TOTAL_CONNECTIONS).evaluateAttributeExpressions().asInteger();
-        final String validationQuery = context.getProperty(VALIDATION_QUERY).evaluateAttributeExpressions().getValue();
-        final Long maxWaitMillis = extractMillisWithInfinite(context.getProperty(MAX_WAIT_TIME).evaluateAttributeExpressions());
-        final Integer minIdle = context.getProperty(MIN_IDLE).evaluateAttributeExpressions().asInteger();
-        final Integer maxIdle = context.getProperty(MAX_IDLE).evaluateAttributeExpressions().asInteger();
-        final Long maxConnLifetimeMillis = extractMillisWithInfinite(context.getProperty(MAX_CONN_LIFETIME).evaluateAttributeExpressions());
-        final Long timeBetweenEvictionRunsMillis = extractMillisWithInfinite(context.getProperty(EVICTION_RUN_PERIOD).evaluateAttributeExpressions());
-        final Long minEvictableIdleTimeMillis = extractMillisWithInfinite(context.getProperty(MIN_EVICTABLE_IDLE_TIME).evaluateAttributeExpressions());
-        final Long softMinEvictableIdleTimeMillis = extractMillisWithInfinite(context.getProperty(SOFT_MIN_EVICTABLE_IDLE_TIME).evaluateAttributeExpressions());
-
+    private void loginKerberos(KerberosUser kerberosUser) throws InitializationException {
         if (kerberosUser != null) {
             try {
                 kerberosUser.login();
             } catch (KerberosLoginException e) {
                 throw new InitializationException("Unable to authenticate Kerberos principal", e);
             }
         }
+    }
 
-        dataSource.setDriver(getDriver(driverName, dburl));
-        dataSource.setMaxWaitMillis(maxWaitMillis);
-        dataSource.setMaxTotal(maxTotal);
-        dataSource.setMinIdle(minIdle);
-        dataSource.setMaxIdle(maxIdle);
-        dataSource.setMaxConnLifetimeMillis(maxConnLifetimeMillis);
-        dataSource.setTimeBetweenEvictionRunsMillis(timeBetweenEvictionRunsMillis);
-        dataSource.setMinEvictableIdleTimeMillis(minEvictableIdleTimeMillis);
-        dataSource.setSoftMinEvictableIdleTimeMillis(softMinEvictableIdleTimeMillis);
+    protected abstract Driver getDriver(final String driverName, final String url);
 
-        if (validationQuery != null && !validationQuery.isEmpty()) {
+    protected DataSourceConfiguration createDataSourceConfiguraton(final ConfigurationContext context) {
+        final String url = context.getProperty(DBCPProperties.DATABASE_URL).getValue();
+        final String user = context.getProperty(DBCPProperties.DB_USER).getValue();
+        final String password = context.getProperty(DBCPProperties.DB_PASSWORD).getValue();
+        final String driverName = context.getProperty(DB_DRIVERNAME).evaluateAttributeExpressions().getValue();
+
+        return new DataSourceConfiguration.Builder(url, driverName, user, password)
+                .build();
+    }
```

Comment:
These four properties are the necessary ones and the optionals are set to default values. I'd assume those won't be modified very often. What do you think about adding a copy constructor to DataSourceConfiguration, so the extended classes could do something like `new DataSourceConfiguration(extendedConfig).Builder().optionalConfigValue1(12)...build()` ?

Reply:
This way we can't override the builder's contructor. Some child classes modifies the url for example which is a required builder field. I think a copy constructor would work.

Reply:
I agree that it is less reusable. The problem with returning with builder is that the mandatory fields can't be changed afterwards. The url is mandatory and is overwritten by several descendant classes. You could override a getUrl method, but I think it would be an overdesign, so I would just make the method plain abstract.